### ISSUE TITLE
[Enhancement] Upgrade: enhance live-migration detection in the pre-drain phase

### ISSUE THREAD

**Issue by bk201** [harvester/harvester#3019]
**Is your feature request related to a problem? Please describe.**

We have trivial rules to detect non-migration VMs in the `upgrade_node.sh` script.
We need to adjust the rule after the v1.1.0 release.

**Describe the solution you'd like**

Some exceptions that we can still live-migrate VMs: 
- We add a mutator to schedule VMs on nodes with a network plane (PR: https://github.com/harvester/harvester/pull/3000).
- The user specifies an affinity that could contain multiple nodes.

We also need to shutdown VMs when passing through PCI devices to VMs.

**Describe alternatives you've considered**

**Additional context**

**Comment by bk201** [comment]
Related: https://github.com/harvester/harvester/issues/2317

**Comment by bk201** [comment]
@starbops Please help backport this to v1.2, it will help reveal  https://github.com/harvester/harvester/issues/5076 in v1.2.1 -> v1.2-head path. Thanks!

**Comment by starbops** [comment]
As there is a problem migrating VMs during the upgrade from v1.2.1 to master (tracked in #5076), I suggest we postpone the testing til the issue is solved.

**Comment by TachunLin** [comment]
Thanks for the checking, according to previous comment, we would move the issue to ready for testing and test til the rke2 have been bumped for the upcoming release.

**Comment by noahgildersleeve** [comment]
Since this is an upgrade test and upgrading two versions of k8s isn't supported should we be running the upgrade from v1.2-head to v1.3.0-rcx for testing? CC @starbops

**Comment by starbops** [comment]
Hi @noahgildersleeve thanks for bringing this up. Yes, we should validate this one by upgrading a cluster from v1.2-head to v1.3.0-rcX. And for the test against the webhook, please try to upgrade from v1.3.0-rcX to v1.3.0-rcX (same version upgrade). Thank you!

**Comment by noahgildersleeve** [comment]
# Test Plan
Full Test plan outlined [here](https://github.com/harvester/harvester/pull/4984) and quoted below
**EDIT: THESE ARE OBSOLETE SCENARIOS AND ARE EXPANDED IN MORE RECENT COMMENT**
- [ ] VMs with node affinities
- [ ] VMs with PCIe devices
- [ ] VMs with node selector

> **Test plan:**
> 
> _VMs with node affinities_
> 
> 1. Prepare a 2-node Harvester cluster in v1.2.1
> 2. Create two VMs: one is live-migratable, the other is non-live-migratable. To achieve this, we could tweak the node scheduling option for the VMs.
>    ![SCR-20240116-okkg](https://private-user-images.githubusercontent.com/1827717/296992499-b8049a2d-4b67-4d2a-afd5-72083014be00.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk4MzY2MjAsIm5iZiI6MTcwOTgzNjMyMCwicGF0aCI6Ii8xODI3NzE3LzI5Njk5MjQ5OS1iODA0OWEyZC00YjY3LTRkMmEtYWZkNS03MjA4MzAxNGJlMDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDdUMTgzMjAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmFiODJiYThlNGY5NTljMDJiMDAwYjkwNGJjZWY0NjY2NDM1NDdmZjRiYmJlZWE3ZDQ4N2Q1Mjg3NWQyM2ZiNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.9UWT10FVlaHNt_m68ppPpPe2jwkfqtUEG8KCRRXpez4)
>    ![SCR-20240116-okdu](https://private-user-images.githubusercontent.com/1827717/296992512-bfb6b721-30b3-4c9b-949e-5215ef6985ce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk4MzY2MjAsIm5iZiI6MTcwOTgzNjMyMCwicGF0aCI6Ii8xODI3NzE3LzI5Njk5MjUxMi1iZmI2YjcyMS0zMGIzLTRjOWItOTQ5ZS01MjE1ZWY2OTg1Y2UucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDdUMTgzMjAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDE4MjEwNzE5NTlkZmE5MWY4OGRiNzYyZjNhMjc1MjFjNzk4MDU5YWI0ZWVmZjM1MWNlMGE0NGUxYTc2OWY3YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.tp8Iurx63oz1EdQiyxP34CTDfrGW3Uo5IdUsit-54cA)
> 3. Upgrade using the master ISO image with any custom-built image which includes this PR
> 4. Monitor the progress of the upgrade, the live-migratable VM should move between the two nodes and the non-live-migratable VM originally running on the control plane node should be shutoff when the control plane node is pre-draining.
> 5. The upgrade should end in success
> 
> _VMs with PCIe devices_
> 
> 1. Prepare a 2-node Harvester cluster in v1.2.1
> 2. Enable `pcidevices-controller` addon
> 3. Enable the device you want for passthrough
> 4. Create a live-migratable VM with regular configs
> 5. Create another non-live-migratable VM with the PCI device attached
> 6. Start the upgrade
> 7. The non-live-migratable VM should be shutoff when the node it runs on is being pre-drained
> 8. The upgrade should end in success and the live-migratable VM is still running
> 
> _VMs with node selector_
> 
> 1. Prepare a 2-node Harvester cluster in v1.2.1
> 2. Create a live-migratable VM with regular configs
> 3. Create another non-live-migratable VM with the node scheduling option set to a specific node
>    ![SCR-20240116-phzu](https://private-user-images.githubusercontent.com/1827717/296997011-db58c1b8-b47b-4482-8ab5-ae646f930cc4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk4MzY2MjAsIm5iZiI6MTcwOTgzNjMyMCwicGF0aCI6Ii8xODI3NzE3LzI5Njk5NzAxMS1kYjU4YzFiOC1iNDdiLTQ0ODItOGFiNS1hZTY0NmY5MzBjYzQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDdUMTgzMjAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Yzc4OTY2Zjk5YTE1MTY5ZDAzZWNmMjUyYzI2NjNjNzk3YjM0ZGEwNWE2OGIyZGQ4YmY4ODdiZmU0ZGNlM2YzNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.xnH8LrRs9t8w-qubjXmN5GChgtpKDYqvMFmkzEPXjQI)
> 4. Start the upgrade
> 5. The VM should be shutoff when the node it runs on is being pre-drained
> 6. The upgrade should end in success and the live-migratable VM is still running
> 
> For the tests against the webhook, the above three test cases could be reused. The only difference is the cluster should be installed with the master build (or custom build containing this PR), not v1.2.1. The upgrade in all three cases should be rejected by the webhook with the names of the VM which are not live-migratable.
> 
> <img alt="image" width="1912" src="https://private-user-images.githubusercontent.com/1827717/297440552-eab51473-e5c0-477d-abe7-ab2c482696ca.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk4MzY2MjAsIm5iZiI6MTcwOTgzNjMyMCwicGF0aCI6Ii8xODI3NzE3LzI5NzQ0MDU1Mi1lYWI1MTQ3My1lNWMwLTQ3N2QtYWJlNy1hYjJjNDgyNjk2Y2EucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDdUMTgzMjAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTVkZDUwNzY2YjA0ZDg1YzU3ZTIzNmM3NDFiN2FhZGZhNmU0Y2NlNDRjMGY1ZmI4MGY2ODFjNGRiYWI0MTNlZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.A68w7hQZbe1QC8lK_-0Oo2HFpkwLVYcy2bmL7MwML_A">

**Comment by noahgildersleeve** [comment]
Tested with v1.2-head -> v1.3.0-rc4 and v1.3.0-rc4 -> v1.3-head/master. I got errors from the webhook on both scenarios. This was on a 2 node vagrant-libvirt cluster using the above instructions for the node affinity example. I can upload both screenshots, but they look the same since I had to manually edit the deployment to look like v1.0.3 for the minupgradeable [workaround](https://github.com/harvester/harvester/issues/3040#issuecomment-1298057127). 
![Screenshot_20240311_141608](https://github.com/harvester/harvester/assets/83787952/cd18a131-3c6f-46a5-aff9-4f516b49e0a5)

![Screenshot_20240311_141333](https://github.com/harvester/harvester/assets/83787952/4bcc2f86-2f38-494c-aef0-d0f66991d340)

**Comment by noahgildersleeve** [comment]
I talked with @starbops and got more clarification. Moving back to testing and I will update the test steps accordingly.

**Comment by starbops** [comment]
The enhancement has been merged into the v1.2 and v1.3 branches. Here I list all possible upgrade paths at the moment and their expected behaviors respectively.

Assuming there is a non-migratable VM running in the cluster:

1. Upgrading from v1.2.1 to v1.2-head will **not** be blocked by the webhhook. However, the VM will be shut down during the node-upgrade phase.
2. Upgrading from v1.2-head to v1.3.0-rc4 will be blocked by the webhook and cannot proceed with the upgrade until the VM is manually shut down
3. Upgrading from v1.3.0-rc1 to v1.3.0-rc4 will **not** be blocked by the webhook, same as 1
4. Upgrading from v1.3.0-rc4 to v1.3.0-rc4 (same version upgrade) will be blocked by the webhook, same as 2.

**Comment by noahgildersleeve** [comment]
New test scenarios for testing. Steps outlined above, but versions are changing

- [x]  VMs with node affinities
   - [x] v1.2.1 -> v1.2-head (VM turns off) PASS issue with #5350 on other VM
   - [x] v1.2-head -> v1.3.0-rc4 (webhook says to turn off VM) PASS
- [ ]  VMs with PCIe devices
   - [ ] v1.2.1 -> v1.2-head (VM turns off)
   - [ ] v1.2-head -> v1.3.0-rc4 (webhook says to turn off VM)
- [x]  VMs with node selector PASS
   - [x] v1.2.1 -> v1.2-head (VM turns off) PASS
   - [x] v1.2-head -> v1.3.0-rc4 (webhook says to turn off VM) PASS

**Comment by noahgildersleeve** [comment]
Closing as validated. We couldn't test with PCI devices due to some environmental issues and #4806 blocking from using a NIC as a passthrough device. All of the other scenarios were tested though.
